crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs

```diff
@@ -304,3 +295,105 @@ pub(crate) fn filter_contained(edits: Vec<Edit>) -> Vec<Edit> {
     }
     filtered
 }
+
+#[derive(Copy, PartialEq, Clone)]
+enum State {
+    Literal,
+    AnnotatedFirst,
+    AnnotatedRest,
+    Other,
+}
+
+pub(crate) struct QuoteAnnotation<'a> {
+    state: Vec<State>,
+    stylist: &'a Stylist<'a>,
+    annotation: String,
+    final_quote_type: Quote,
+}
+
+impl<'a> QuoteAnnotation<'a> {
+    pub(crate) fn new(stylist: &'a Stylist<'a>) -> Self {
+        let final_quote_type = stylist.quote();
+        Self {
+            state: vec![],
+            stylist,
+            annotation: String::new(),
+            final_quote_type,
+        }
+    }
+}
+
+impl<'a> source_order::SourceOrderVisitor<'a> for QuoteAnnotation<'a> {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        let generator = Generator::from(self.stylist);
+        match expr {
+            Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+                if let Some(name) = value.as_name_expr() {
+                    let value = generator.expr(value);
+                    self.annotation.push_str(&value);
+                    self.annotation.push('[');
+                    match name.id.as_str() {
+                        "Literal" => self.state.push(State::Literal),
+                        "Annotated" => self.state.push(State::AnnotatedFirst),
+                        _ => self.state.push(State::Other),
+                    }
+
+                    self.visit_expr(slice);
+                    self.state.pop();
+                    self.annotation.push(']');
+                }
+            }
+            Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+                let Some(first_elm) = elts.first() else {
+                    return;
+                };
+                self.visit_expr(first_elm);
+                if self.state.last().copied() == Some(State::AnnotatedFirst) {
+                    self.state.push(State::AnnotatedRest);
+                }
+                for elm in elts.iter().skip(1) {
+                    self.annotation.push_str(", ");
+                    self.visit_expr(elm);
+                }
+                self.state.pop();
+            }
+            Expr::BinOp(ast::ExprBinOp {
+                left, op, right, ..
+            }) => {
```

Comment:
I think we're only interested in a binary or operation here, right? And, I think at this stage we'll only be getting them because others are filtered out in `quote_annotation` function above. But, it might make sense to still enforce it here by using a `debug_assert`.

Reply:
We should be right. Added it.